x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -678,6 +685,47 @@ private List<Attribute> resolveUsingColumns(List<Attribute> cols, List<Attribute
             return resolved;
         }
 
+        private LogicalPlan resolveInsist(Insist insist, List<Attribute> childrenOutput, IndexResolution indexResolution) {
+            Attribute resolvedCol = maybeResolveAttribute(insist.getInsistIdentifier(), childrenOutput);
+            // Field isn't mapped anywhere.
+            if (resolvedCol instanceof UnresolvedAttribute) {
+                return pushdownInsist(insist, attrs -> attrs.add(insistKeyword(insist)));
+            }
+
+            String name = resolvedCol.name();
+            // Field is partially unmapped.
+            if (resolvedCol instanceof FieldAttribute f && indexResolution.get().partiallyUnmappedFields().contains(name)) {
+                return pushdownInsist(insist, attrs -> {
+                    var index = CollectionUtils.findFirstIndex(attrs, e -> e.name().equals(name)).getAsInt();
+                    Attribute attribute = f.field().getDataType() == KEYWORD ? insistKeyword(insist) : invalidInsistAttribute(insist, f);
+                    attrs.set(index, attribute);
+                });
+            }
+
+            // Field is mapped everywhere; we can safely ignore the INSIST command.
+            return insist.child();
+        }
+
+        private static EsRelation pushdownInsist(Insist insist, Consumer<List<Attribute>> updateAttributes) {
+            var relation = (EsRelation) insist.child();
+            List<Attribute> newOutput = new ArrayList<>(relation.output());
+            updateAttributes.accept(newOutput);
+            return relation.withAttributes(newOutput);
+        }
```

Comment:
Create a separate rule in the optimizer that picks up Project with unmappedEsField & co which folds (moves) these fields into the source.